src/ferui-components/tree-view/interfaces.ts

```diff
@@ -0,0 +1,108 @@
+import { TemplateRef } from '@angular/core';
+
+/**
+ * Tree Node Data Interface
+ */
+export interface TreeNodeData<T> {
+  data: T;
+  label: string;
+}
+
+/**
+ * Paging Params Interface
+ */
+export interface PagingParams {
+  offset: number;
+  limit: number;
+}
+
+/**
+ * Tree Node Data Retriever Interface for client-side tree view
+ */
+export interface TreeNodeDataRetriever<T> {
+  // Get promise that returns TreeNodeData children array from Tree Node
+  getChildNodeData(parent: TreeNode<T>): Promise<Array<TreeNodeData<T>>>;
+  // Get promise that returns whether TreeNode has children nodes
+  hasChildNodes(parent: TreeNode<T>): Promise<boolean>;
+  // based on the current TreeNode and its status, we allow the developer to give us an icon template
+  getIconTemplate?(treeNode: TreeNode<T>): TemplateRef<any>;
+  // If developer wishes to give us a template ref we shall render this template in the view
+  getNodeTemplate?(): TemplateRef<any>;
```

Comment:
why do we need both getNodeTemplate and getIconTemplate.. doesn't getNodeTemplate also permit the user to change the icon? How does it currently work when both are specified?

Reply:
Actually the `getIconTemplate()` control how the triggering icon is shown (the arrows in our examples) and `getNodeTemplate()` method control the node "label". 
We can't control both in the same template since @mathieupkf want to have ONLY the icon to be clickable for expanding/collapsing children. If we put everything in the same template we won't be able to do it.

Reply:
The getIconTemplate is for the icons to the left of the node and we use these to bind to our expand and collapse methods. Otherwise we would have to bind “this” to all templates given and depend on the developer to do this for us, so instead we just ask for an icon and we do the rest. 
The getNodeTemplate they can do whatever they want with and will only display the view how they want without needing anything extra.

Reply:
I think it would be more flexible to just let the developer control both the icon and the label in a single template. Of course, our default template should follow Mathieu's guidelines in terms of clicking behavior, but there is no reason to prevent others from changing that behavior if they prefer to use a custom template.

Reply:
Well, if we decide to go with only one template, we would need to bind the clicking event through ng-template, and as a dev, if we just want to change the label template, we would also need to remember that we need to add manually the arrow icon and bind the clicking event to it... 
With separated templates like we did, we could easily change only one part of the node vs always changing the whole node (including label + arrow icon).
The only thing that we can't control with having two templates is the position of the icon template: it will always be a the left from the label.